text/2018-11-24-consistent-replica-read.md

```diff
@@ -0,0 +1,67 @@
+# Summary
+`Consistent Replica Read` supports consistent read on the follower replica or 
+learner replica, which can reduce the read load on the leader. `Consistent 
+Replica Read` also useful in OLAP scenario or backup scenario.
+
+The key to achieve consistent read on the replica is `ReadIndex`, which 
+introduced in chapter 6 of the (raft 
+thesis)[https://ramcloud.stanford.edu/~ongaro/thesis.pdf].
+
+# Motivation
+
+Currently, All read requests are handled on the leader, load balance only by 
+`balance leader` operator, there are lots of limitations. On one hand, we hope 
+a way to improve read throughput and divert load away from the leader. and On 
+the other hand, hope some nodes support read-only service and handle some heavy 
+analytics queries.
+
+# Detail design
+
+## Consistent Replica Read
+As summary said, the key to implementing `Consistent Replica Read` is 
+`ReadIndex`, which already implemented in our 
+(raft-rs)[http://github.com/pingcap/raft-rs] library. For now, some read 
+requests already use `ReadIndex` policy to handle, as the following steps.
+
+1. Determine the policy of the read request. If the request is required to 
+check the quorum, the replica is a leader and expired the lease, or the replica 
+is a follower or a learner, should use `ReadIndex` policy.
+2. Got `ReadIndex` from the raft library. Assign a `uuid` for the request and 
+invoke the `read_index` API of the raft library with the data that encoded by 
+`uuid`. then the raft state machine will notify our apply state machine with 
+the `read_status` through the `Ready` message. and we can get the `ReadIndex` 
+and `uuid` from the `read_status`.
+3.  Read data according to the state of applying. If the `apply_index_term` 
+equal to the `current_term`, the request can be processed, rather than will 
+push to the pending queue.
+**Note**: For the leader, the request can be processed if the 
+`apply_index_term` equal to the `current_term`. For the follower and the 
+learner, should let `apply_index` greater than or equal to the `ReadIndex`.
```

Comment:
hi, @nolouch , I am a little bit confusing here. Because i think for ensuring linearizability read, even for the leader, we still need wait  apply_index greater than or equal to the commit_index. Why we can handle read request when apply_index_term equals to the current_term?

Reply:
In TiKV, the write request is successful after the `apply` is finished. If the write request actually succeeded in that term, the `apply_index` will greater than or equal to the write entry index. the read request after it will read the newly written data.